### ISSUE TITLE
[BUG] SFTLoss and DistillationLoss cache self.device and break after .to()

### ISSUE THREAD

**Issue by YeonwooSung** [pytorch/rl#4224]
## Describe the bug

`SFTLoss` and `DistillationLoss` both assign `self.device = device` on a `LossModule` / `nn.Module`. `module.to(...)` moves parameters and buffers only, so the cached Python attribute goes stale. After `loss.to("cuda")` (or `.cpu()`), mask recovery and the actor forward still use the constructor device (often `None` or `"cpu"`).

This violates the TorchRL module-device contract (Agents.md §6a): do not cache a module device; derive placement from the involved tensor.

## To Reproduce

Both classes store the constructor argument as module state:

```259:298:torchrl/objectives/llm/sft.py
    def __init__(
        self,
        ...
        device: torch.device | None = None,
    ):
        super().__init__()
        ...
        self.device = device
```

```307:345:torchrl/objectives/llm/distillation.py
    def __init__(
        self,
        ...
        device: torch.device | None = None,
    ):
        super().__init__()
        ...
        self.device = device
```

Forward then wraps tokenizer / actor work in `torch.device(self.device)`:

```345:376:torchrl/objectives/llm/sft.py
        if assistant_masks is None:
            with torch.device(
                self.device
            ) if self.device is not None else contextlib.nullcontext():
                token_struct = history.apply_chat_template(
                    tokenizer=self.tokenizer, **self.tokenizer_kwargs
                )
        ...
        with torch.device(
            self.device
        ) if self.device is not None else contextlib.nullcontext():
            output_loss = self.actor_network(input_loss)
```

The same pattern is in `DistillationLoss._get_masks` (`distillation.py:365-370`) and `DistillationLoss.forward` (`distillation.py:439-442`).

Sketch:

```python
loss = SFTLoss(actor_network=actor, tokenizer=tokenizer, device="cpu")
loss.to("cuda")          # moves parameters/buffers; self.device stays "cpu"
out = loss(td.cuda())    # mask/token work still runs under torch.device("cpu")
```

The same applies to `DistillationLoss`.

## Expected behavior

After `loss.to("cuda")` or `loss.to("cpu")`, mask recovery and the actor forward should use the device of the input tensordict or of a parameter / buffer, not a cached constructor string. Dropping `self.device` and deriving the device from the involved tensor is the contract used elsewhere in TorchRL.

## Screenshots

N/A

## System info

Observed on a source checkout of pytorch/rl at `1d3de3db`.

## Additional context

`LossModule` is an `nn.Module`. A module may also span several devices, in which case no single `self.device` is valid. Overriding `to()` / `_apply()` to keep the cache in sync is not an acceptable workaround (same contract).

## Reason and Possible fixes

Delete the `self.device` attribute. When a device context is still needed for tokenizer output, take it from `tensordict.device`, from `next(self.parameters(), None)`, or from the first mask / log-prob tensor already on the module. Do not re-introduce a cached module-level device.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)
